README.md
<h4 align="right">English | <strong><a href="https://github.com/LTopx/Le-AI/blob/main/README_CN.md">中文</a></strong></h4>

<p align="center">
    <a href="https://le-ai.app" target="_blank" rel="noopener noreferrer">
        <img width="100" src="./public/favicon-96x96.png" alt="Le-AI" />
    </a>
</p>

<h1 align="center">Le-AI</h1>

<p align="center">Your open-source AI Assistant Hub, helping you boost efficiency UP UP~</p>

<p align="center">
<a href="https://github.com/LTopx/Le-AI/releases">
  <img alt="GitHub release" src="https://img.shields.io/github/release/LTopx/Le-AI.svg?style=flat-square&include_prereleases" />
</a>
<a href="https://github.com/LTopx/Le-AI/commits">
  <img alt="GitHub last commit" src="https://img.shields.io/github/last-commit/LTopx/Le-AI.svg?style=flat-square" />
</a>
<a href="https://twitter.com/peekbomb" target="_blank">
  <img alt="twitter" src="https://img.shields.io/badge/follow-Ethan Liu-red?style=flat-square&logo=Twitter">
</a>
<a href="https://t.me/+7fLJJoGV_bJhYTk1" target="_blank">
  <img alt="telegram" src="https://img.shields.io/badge/chat-telegram-blueviolet?style=flat-square&logo=Telegram">
</a>

<div align="center">

[Demo](https://le-ai.app/) | [Docs](https://docs.le-ai.app/) | [Q&A](https://docs.le-ai.app/faq) | [Change Log](https://docs.le-ai.app/change-log) | [Feedback](https://github.com/LTopx/Le-AI/issues) | [Telegram](https://t.me/+7fLJJoGV_bJhYTk1) | [Contact Me](https://goethan.cc/)

</div>

<div align="center">

[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/new/clone?repository-url=https://github.com/LTopx/Le-AI)

</div>

![cover](./public/screenshots/screenshot.png)

## ✨ Demo

Direct access: [https://le-ai.app](https://le-ai.app/)

Project documentation: [https://docs.le-ai.app](https://docs.le-ai.app/)

## 🎯 Key Features

- No need to configure additional environment variables, can be easily deployed to Vercel for free
- Ensures privacy and security, all session records and system configurations are stored locally in the browser
- Responsive design with dark mode, providing a great experience on different devices
- Supports voice reading with customizable voices and speeds
- Supports displaying markdown with code highlighting and copy operations
- Supports OpenAI and Azure OpenAI
- Supports custom role templates to create more AI possibilities
- Supports i18n multilingual internationalization: English, Simplified Chinese
- Support Docker deployment
- For more information, please refer to the [documentation](https://docs.le-ai.app/)

## 📍 Development Plan

- [x] Support custom prompt repository
- [x] Support Function call for implementing more functionalities
- [x] Support for large language model APIs such as Claude, PaLM, and Llama 2
- [x] Support distributing keys for use in an unauthenticated state
- [x] Support unlimited sessions
- [ ] Support integration of Midjourney drawing
- [ ] Desktop version development

## 💿 Deployment

### Docker Deployment (Recommended)

```
docker pull ltopx/le-ai:latest

docker run -d -p 3000:3000 ltopx/le-ai:latest
```

### Docker Local Deployment

```
docker build -t ltopx/le-ai .

docker run -d -p 3000:3000 ltopx/le-ai
```

### One-click Deployment

Currently supports one-click deployment to Vercel.
[![Deploy with Vercel](https://vercel.com/button)](https://vercel.com/new/clone?repository-url=https://github.com/LTopx/Le-AI)

## 🪄 Local Development

**0. Node Environment Requirements**

NodeJS >= 18

**1. Install PNPM**

If you have not previously installed or used `pnpm`, you can install it by running the following command.

```bash
npm install pnpm -g
```

**2. Install Dependencies**

```bash
pnpm i
```

**3. Configure Environment Variables**

Rename .evn.local.demo to .env.local

**4. Run the Project**

```bash
pnpm dev
```

**5. Build the Project**

```bash
pnpm build && pnpm start
```

## More Optional Environment Variables

Refer to the documentation: [https://docs.le-ai.app](https://docs.le-ai.app/develop/env)

## License

[GNU](https://github.com/LTopx/Le-AI/blob/main/LICENSE)


CHANGE_LOG.md
# Le-AI Change Log

## v0.9.3

> 2023-09-27

### Fixed

- Fix the issue where the pop-up window under the avatar dropdown menu cannot be displayed properly.
- Fix the issue where metadata image cannot be displayed properly.

## v0.9.2

> 2023-09-27

### Add

- Support Docker deployment, [#15](https://github.com/LTopx/Le-AI/pull/15)
- Added support for configuring a member threshold for session summarization to achieve the effect of "unlimited sessions"
- "Session Settings" added support for configuring the threshold of message compression for historical messages.

### Fixed

- Fix avatar UI display error issue
- Fix configuration error in .env.local.demo file
- Fix multiple language configuration errors
- Fix the issue of parsing error in react-markdown caused by upgrading remark-gfm version
- Fix the display effect of the loading icon during session loading
- Fix the error of missing default time zone for next-intl
- Fix the issue where the system settings menu cannot be opened properly in some cases
- Fixed [#10](https://github.com/LTopx/Le-AI/issues/10)

### Changed

- Reconstruct some page module UI and migrate to shadcn/ui
- Adjust the proxy configuration logic for OpenAI API Key, fix [#11](https://github.com/LTopx/Le-AI/issues/11)
- Disable voice playback during conversation loading.
- Adjust the timeout time of Serverless Functions to 5 minutes.
- Adjust account center UI interface.

## v0.9.1

> 2023-09-11

### Fixed

- Fixed the issue of incorrect authentication logic after configuring a custom API Key when deploying a project
- Fixed the issue of garbled characters in some cases when replying
- Fixed [#8](https://github.com/LTopx/Le-AI/issues/8)
- Fixed the issue of incorrect background image display on the email login page
- Fix the issue of character creation prompt and welcome message word limit

### Add

- Support for account/password login
- Added a return to home button on the API Key configuration page
- Added privacy agreement and login guide to the login interface.
- Compatible with configuring other third-party OpenAI API Keys. For example: nextweb, API2D, and so on.

### Changed

- Update Azure OpenAI API Version to 2023-08-01-preview
- Adjust the UI of the left bottom menu in the system
- Refactor the UI of the pop-up box for creating a new character
- Adjust the implementation logic of prompt, messages parameters, all parsed and assembled by the front end

## v0.9.0

> 2023-09-02

### Fixed

- Fixed the issue of being unable to use when not logged in but configured with an API KEY
- Fixing invalid links.
- Fixing the cross-domain issue with the chatgpt interface.
- Fixing the issue of missing metadataBase.
- Fixed the redirect link issue for OpenRouter Key.
- Fixing the issue of not carrying cross-domain headers when requesting an error.
- Fixed the style issue with the Table on the token page.
- Fixed: issue of missing metadataBase and env variable [#6](https://github.com/LTopx/Le-AI/issues/6)

### Add

- Added support for Windows and Mac Beta clients.
- Added support for one-click session clearing
- Added logic and database table construction for token management page
- Added important announcement prompt in the left-side menu
- Self-built token integration logic for OpenAI, Azure OpenAI, and OpenRouter interfaces
- Added support for configuring Le-AI API Key
- Introduced shadcn/ui to optimize the style of some components
- Added support for Traditional Chinese

### Changed

- Adjust the project name: L-GPT => Le-AI
- Adjust all the copy and related links within the project regarding L-GPT.
- Adjusted some code structure
- Add the Le-AI API Key to the logic for determining whether the key is configured.
- Adjusted the user credential expiration time to 14 days.
- Refactor and streamline global multilingual configuration.
- Refactor the login interface UI.
- Adjusted the permissions of some models in OpenRouter.

## v0.8.4

> 2023-08-19

### Add

- Support for configuring welcome messages has been added to the character template, serving as the default first sentence of a conversation.
- Add privacy policy and contact information.
- Added cloud synchronization function, which can synchronize conversation content to the cloud for easy synchronization between different devices.

### Changed

- Support for importing and exporting configurations of OpenRouter.
- Adjust the layout of the bottom left menu.
- Adjust the model permissions provided by OpenRouter to free.
- Adjust the cloud backup quota, free users can back up 50KB of data.

## v0.8.3

> 2023-08-13

### Fixed

- Fixed mobile session content obscuring the bottom input box
- Refactored function calling invocation logic and fixed bugs
- Fixed the "drift" of the drop-down selection box when selecting a new session model
- Fix the issue of squeezed avatars in the conversation list.

### Add

- Added function calling support
- Added plugin system
- Added support for Google search, which can call the Google API to search and return results when encountering problems that exceed the AI model training date
- Introduced OpenRouter to support Claude, PaLM2, Llama 2 and other models

### Changed

- Adjusted the text input box for editing chat content to Textarea
- Replaced Google search with [Serper API](https://serper.dev/), which is easier to configure
- All models use openai gpt-3.5-turbo to get conversation titles, saving token consumption
- When using the models provided by OpenRouter, the plugins are hidden because they do not support plugins at this time

## v0.8.2

> 2023-08-07

### Fixed

- Fixed the problem that AUTH_SECRET may be missing in the production environment

### Add

- Added support for Azure GPT-4/GPT-4-32K
- Added error message prompt when NEXT_PUBLIC_AZURE_OPENAI_RESOURCE_NAME is not configured

### Changed

- Adjust the environment variable configuration and related code logic, now you can deploy with one click without any environment variables
- Adjust readme documentation

a.js
addEventListener("fetch", (event) => {
  event.respondWith(handleRequest(event.request));
});

const config = {
  openai: {
    originUrl: "https://api.openai.com",
  },
  azure: {
    apiVersion: "2023-03-15-preview",
    // The name of your Azure OpenAI Resource.
    resourceName: "lgpt-azure-openai",
    // The deployment name you chose when you deployed the model.
    model: {
      "gpt-3.5-turbo-0301": "lgpt-35-turbo",
    },
  },
};

async function handleRequest(request) {
  if (
    !request.url.includes("openai-proxy") &&
    !request.url.includes("azure-proxy")
  ) {
    return new Response("404 Not Found", { status: 404 });
  }

  if (request.method === "OPTIONS") return handleOPTIONS(request);

  if (request.url.includes("openai-proxy")) {
    return handleOpenAI(request);
  } else if (request.url.includes("azure-proxy")) {
    return handleAzure(request);
  }
}

async function handleOpenAI(request) {
  const url = new URL(request.url);

  url.host = config.openai.originUrl.replace(/^https?:\/\//, "");

  const modifiedRequest = new Request(url.toString(), {
    headers: request.headers,
    method: request.method,
    body: request.body,
    redirect: "follow",
  });

  const response = await fetch(modifiedRequest);
  const modifiedResponse = new Response(response.body, response);

  // 添加允许跨域访问的响应头
  modifiedResponse.headers.set("Access-Control-Allow-Origin", "*");

  return modifiedResponse;
}

async function handleAzure(request) {
  const url = new URL(request.url);

  if (url.pathname === "/v1/chat/completions") {
    var path = "chat/completions";
  } else if (url.pathname === "/v1/completions") {
    var path = "completions";
  } else if (url.pathname === "/v1/models") {
    return handleModels(request);
  } else {
    return new Response("404 Not Found", { status: 404 });
  }

  let body;
  if (request.method === "POST") body = await request.json();

  const modelName = body?.model;
  const deployName = config.azure.model[modelName];

  if (deployName === "") return new Response("Missing model", { status: 403 });

  const fetchAPI = `https://${config.azure.resourceName}.openai.azure.com/openai/deployments/${deployName}/${path}?api-version=${config.azure.apiVersion}`;

  const authKey = request.headers.get("Authorization");

  if (!authKey) return new Response("Not allowed", { status: 403 });

  const payload = {
    method: request.method,
    headers: {
      "Content-Type": "application/json",
      "api-key": authKey.replace("Bearer ", ""),
    },
    body: typeof body === "object" ? JSON.stringify(body) : "{}",
  };

  const response = await fetch(fetchAPI, payload);

  if (body?.stream != true) {
    return response;
  }

  const { readable, writable } = new TransformStream();
  stream(response.body, writable);
  return new Response(readable, response);
}

function sleep(ms) {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

// support printer mode and add newline
async function stream(readable, writable) {
  const reader = readable.getReader();
  const writer = writable.getWriter();

  // const decoder = new TextDecoder();
  const encoder = new TextEncoder();
  const decoder = new TextDecoder();
  // let decodedValue = decoder.decode(value);
  const newline = "\n";
  const delimiter = "\n\n";
  const encodedNewline = encoder.encode(newline);

  let buffer = "";
  while (true) {
    let { value, done } = await reader.read();
    if (done) {
      break;
    }
    buffer += decoder.decode(value, { stream: true }); // stream: true is important here,fix the bug of incomplete line
    let lines = buffer.split(delimiter);

    // Loop through all but the last line, which may be incomplete.
    for (let i = 0; i < lines.length - 1; i++) {
      await writer.write(encoder.encode(lines[i] + delimiter));
      await sleep(30);
    }

    buffer = lines[lines.length - 1];
  }

  if (buffer) {
    await writer.write(encoder.encode(buffer));
  }
  await writer.write(encodedNewline);
  await writer.close();
}

async function handleModels() {
  const data = {
    object: "list",
    data: [],
  };

  for (let key in config.azure.model) {
    data.data.push({
      id: key,
      object: "model",
      created: 1677610602,
      owned_by: "openai",
      permission: [
        {
          id: "modelperm-M56FXnG1AsIr3SXq8BYPvXJA",
          object: "model_permission",
          created: 1679602088,
          allow_create_engine: false,
          allow_sampling: true,
          allow_logprobs: true,
          allow_search_indices: false,
          allow_view: true,
          allow_fine_tuning: false,
          organization: "*",
          group: null,
          is_blocking: false,
        },
      ],
      root: key,
      parent: null,
    });
  }

  const json = JSON.stringify(data, null, 2);
  return new Response(json, {
    headers: { "Content-Type": "application/json" },
  });
}

async function handleOPTIONS() {
  return new Response(null, {
    headers: {
      "Access-Control-Allow-Origin": "*",
      "Access-Control-Allow-Methods": "*",
      "Access-Control-Allow-Headers": "*",
    },
  });
}
